File: skyscrapers.pl
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% File: skyscrapers.pl
% Brief: Implementazione del gioco skyscrapers
% Author: Francesco Vetere
% Date: 2021/06/21
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%% Esempio d'uso %%%%%%%%%%%%%%%%%
% swipl
%
% ?- consult('skyscrapers.pl').
%
% ?- isMatrixCompliant([[0,0,0,0,0],[0,1,2,3,1],[0,3,1,2,0],[2,2,3,1,0],[0,0,0,3,0]]). 
%		==> true.
%
% ?- isMatrixCompliant([[0, 3, 2, 1, 3, 2, 0],[3, 3, 4, 5, 1, 2, 2],[4, 1, 2, 4, 3, 5, 1],[3, 2, 1, 3, 5, 4, 2],[2, 4, 5, 1, 2, 3, 2],[1, 5, 3, 2, 4, 1, 3],[0,1, 2, 4, 2, 4, 0]]). 
% 	==> true.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%% Predicati di utilita' su liste %%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


% member(elem, L) ==> true se L contiene elem
member(X, [X | _]).
member(X, [_ | T]) :- member(X, T).

% append(L, R, RIS) ==> RIS è la lista L concatenata con R
append([], R, R).
append([H | T], R, [H | TEMP]) :- append(T, R, TEMP).

% reverse(A, B) ==> B è la lista A rovesciata
reverse([], []).
reverse([H | T], RIS) :- reverse(T, TEMP), append(TEMP, [H], RIS).

% length(A, LEN) ==> LEN è pari alla lunghezza della lista A
len([], 0).
len([_ | T], RIS) :- len(T, TEMP), RIS is TEMP + 1.

% removeFirstLast(A, RIS) ==> RIS è la lista A privata del primo e dell'ultimo elemento
removeFirstLast(L, RIS) :- removeFirst(L, TEMP), removeLast(TEMP, RIS).

% removeFirst(A, RIS) ==> RIS è la lista A privata del primo elemento
removeFirst([], []).
removeFirst([_], []).
removeFirst([_ | T], T). 

% removeLast(A, RIS) ==> RIS è la lista A privata dell'ultimo elemento
removeLast([], []).
removeLast([_], []).
removeLast(X, Y) :- reverse(X, [_ | T]), reverse(T, Y).

% unique(L) ==> true se L contiene valori unici
unique([]).
unique([H | T]) :- \+ member(H, T), unique(T).

% uniqueFirstLast(L) ==> true se L, privata del primo e ultimo elemento (quelli di bordo!) contiene valori unici
uniqueFirstLast(L) :- removeFirstLast(L, LNew), unique(LNew).


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%% Predicati di utilita' su matrici %%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


% uniqueMatrixFirstLast(M) ==> true se la matrice M contiene liste uniche, private del primo e ultimo elemento (quelli di bordo!)
uniqueMatrixFirstLast([]).
uniqueMatrixFirstLast([H | T]) :- 
		uniqueFirstLast(H), uniqueMatrixFirstLast(T).

% isMatrixSquare(M) ==> true se M è almeno 2x2 ed è quadrata
isMatrixSquare([H | T]) :- len(H, COLS), len([H | T], ROWS), 
													 ROWS >= 2, COLS >= 2, ROWS == COLS.

% transposeMatrix(M, MT) ==> MT contiene la trasposta di M
transposeMatrix([[] | _], []).
transposeMatrix(Matrix, [FirstColumn | TransposedRestMatrix]) :- 
		splitMatrix(Matrix, FirstColumn, RestMatrix), % FirstColumn diventa la prima riga nella matrice trasposta
    transposeMatrix(RestMatrix, TransposedRestMatrix).

% splitMatrix(Matrix, FirstColumn, RestOfMatrix) ==> Matrix viene divisa in FirstColumn e RestOfMatrix
splitMatrix([], [], []).
splitMatrix(
		[[FirstEl | RestFirstRow] | OtherRows],   							% Matrix
		[FirstEl | RestFirstCol],  														 	% First column
		[RestFirstRow | RestOtherRows]												 	% Rest of the matrix
) :- splitMatrix(OtherRows, RestFirstCol, RestOtherRows).


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%% Predicati specifici di skyscrapers %%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


% countChangesMax(L, RIS) ==> RIS è pari al numero di volte in cui il massimo cambia nella lista L (da sinistra a destra) 
countChangesMax([0 | T], RIS) :- countChangesMaxAux(T, 0, 1, RIS).
countChangesMax(L, RIS) :- countChangesMaxAux(L, 0, 0, RIS).

% countChangesMaxAux(L, RIS) ==> RIS è pari al numero di volte in cui il massimo cambia nella lista L (da sinistra a destra)
% 															 con massimo corrente pari a CurrentMax e accumulatore Acc
countChangesMaxAux([H | T], CurrentMax, Acc, RIS) :-
    H > CurrentMax,
		AccNew is Acc + 1,
    countChangesMaxAux(T, H, AccNew, RIS).
 
countChangesMaxAux([H | T], CurrentMax, Acc, RIS) :-
    H =< CurrentMax,
    countChangesMaxAux(T, CurrentMax, Acc, RIS).
 
countChangesMaxAux([], _, Acc, Acc).

% isListCompliant(L) ==> true se la lista L rispetta il vincolo del gioco skyscrapers previsto per ogni riga/colonna:
% 											 il primo elemento della lista deve essere uguale al numero di massimi incontrati
% 											 lungo il resto della lista
isListCompliant([0 | _]). 
isListCompliant([H | T]) :-  countChangesMax(T, MAXS), H == MAXS. 

% checkRuleForward(M) ==> true se la matrice M rispetta il vincolo del gioco skyscrapers per ogni riga, da sx a dx
checkRuleForward([]).
checkRuleForward([H | T]) :-
		isListCompliant(H), checkRuleForward(T).

% checkRuleBackward(M) ==> true se la matrice M rispetta il vincolo del gioco skyscrapers per ogni riga, da dx a sx
checkRuleBackward([]).
checkRuleBackward([H | T]) :-
		reverse(H, HReverse), isListCompliant(HReverse), checkRuleBackward(T).

% isMatrixCompliant(L) ==> true se la matrice M rispetta le regole del gioco skyscrapers
isMatrixCompliant(M) :-
		isMatrixSquare(M), 																				% controllo che la matrice sia quadrata

		removeFirstLast(M, M1), uniqueMatrixFirstLast(M1),        % controllo che all'interno dei bordi vi siano valori unici:
																															% elimino prima e ultima riga, e controllo se le restanti righe
																															% contengono valori unici a meno di prima e ultima colonna
		
		checkRuleForward(M),																			% controllo la regola da sx a dx
		checkRuleBackward(M),																			% controllo la regola da dx a sx

		transposeMatrix(M, MT),
		
		checkRuleForward(MT),																			% traspongo e controllo la regola da sx a dx
		checkRuleBackward(MT).																		% traspongo e controllo la regola da dx a sx

File: haskell/skyscrapers.hs
-- File: skyscrapers.hs
-- Brief: Implementazione del gioco skyscrapers
-- Author: Francesco Vetere 
-- Date: 2020/05/09

-- Esempio d'uso
-- > ghc --make skyscrapers.hs
-- > ./skyscrapers
-- > "Insert file name:"
-- > skyscrapers-5x5.txt
-- > "Matrix compliant"

import System.IO
import Data.List  -- reverse, transpose, nub

-- Il main racchiude la parte "impura", cioè non funzionale pura (I/O)
main = do
    print "Insert file name:"
    file_name <- getLine
    file_content <- readFile file_name  -- leggo tutto il file, ottenendo una lunga stringa

    -- divido il contenuto del file per righe separate da '\n' (tramite la funzione lines)
    -- ogni linea e' analizzata tramite una scanl, che mappa la funzione read :: [Int] sulla 
    -- lista formata dalle singole cifre separate da ' ' (tramite la funzione words)
    -- infine, si elimina la prima lista vuota che viene generata, tramite la funzione tail
    -- si ottiene dunque una matrice, che viene passata alla funzione isMatrixCompliant,
    -- e dunque la print ne stampa il risultato finale

    if (isMatrixCompliant . tail . scanl (\acc x -> (map (read) (words x) :: [Int])) [] . lines $ file_content)
    then print "Matrix compliant"
    else 
        print "Matrix not compliant"

-----------------------------------------------------------------------------

-- Data una lista, calcola il numero di volte che il massimo cambia
countChangesMax :: [Int] -> Int
countChangesMax values = length $ foldl (\ acc x -> if null acc || x > head acc then x:acc else acc) [] values

-- Data una lista, restituisce la lista privata del primo e dell'ultimo elemento
removeFirstLast :: [a] -> [a]
removeFirstLast list =
    if (length list >= 2) then tail . init $ list
    else []

-- Data una lista, controlla se, tolti il primo e l'ultimo elemento (che sono quelli di bordo!), contiene valori unici
uniqueValues :: (Eq a) => [a] -> Bool
uniqueValues list =
    let l_ = removeFirstLast list
    in
        if ((length (nub l_)) == length l_) then True
        else False

-- Data una matrice, controlla se essa e' almeno una 2x2, ed e' quadrata
isMatrixSquare :: [[a]] -> Bool
isMatrixSquare m =
    if (
        length m > 2                    -- almeno 2 righe
        &&
        length (m !! 0) > 2             -- almeno 2 colonne
        &&
        length m == length (m !! 0)     -- numero righe = numero colonne
    )
    then True
    else False

-- Data una lista, controlla che rispetti il vincolo del gioco skyscrapers previsto per ogni riga/colonna:
-- il primo elemento della lista deve essere uguale al numero di massimi incontrati lungo il resto della lista
isListCompliant :: [Int] -> Bool
isListCompliant l =
    if (l !! 0 == (countChangesMax $ tail l)) then True 
    else False

-- Data una matrice, verifica se essa rispetta le regole del gioco skyscrapers
isMatrixCompliant :: [[Int]] -> Bool
isMatrixCompliant m =
    if (
        isMatrixSquare m                                                        -- controllo che la matrice sia quadrata
            &&
        (not $ False `elem` (
            map uniqueValues (removeFirstLast m)                                -- controllo che all'interno dei bordi vi siano valori unici
        ))
            &&
        (not $ False `elem` 
        (   map isListCompliant (removeFirstLast m) ++                          -- controllo la regola da sx a dx
                map isListCompliant (reverse (removeFirstLast m)) ++            -- controllo la regola da dx a sx
                map isListCompliant (removeFirstLast (transpose m)) ++          -- traspongo, e controllo la regola da sx a dx
                map isListCompliant (reverse (removeFirstLast (transpose m)))   -- traspongo, e controllo la regola da dx a sx
        ))
    )
    then True
    else False
